tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.hh

```diff
@@ -0,0 +1,337 @@
+    public: const std::vector<double> &AngularFrequency_V() const;
+
+    /// \brief Access the component amplitudes.
+    public: const std::vector<double> &Amplitude_V() const;
+
+    /// \brief Access the component phases.
+    public: const std::vector<double> &Phase_V() const;
+
+    /// \brief Access the steepness components.
+    public: const std::vector<double> &Steepness_V() const;
+
+    /// \brief Access the component wavenumbers.
+    public: const std::vector<double> &Wavenumber_V() const;
+
+    /// \brief Access the component directions.
+    public: const std::vector<gz::math::Vector2d> &Direction_V() const;
+
+    /// \brief Print a summary of the wave parameters to the gzmsg stream.
+    public: void DebugPrint() const;
+
+    /// A simpler version of determining wave height at a point.
+    /// This method enforces that q (steepness) = 0 which allows us
+    /// to caculate the wave height exactly for a given 2D point without the
+    /// need to interatively solve for the position/height.
+    /// \param[in] _point       The point at which we want the depth.
+    /// \param[in] _time        The time at which we want the depth.
+    /// \param[in] _timeInit    The time at which we want the wavefield to start
+    /// \return                 The depth 'h' at the point.
+    public: double ComputeDepthSimply(const gz::math::Vector3d &_point,
+                                      double _time,
+                                      double _timeInit = 0);
+
+    /// \brief Compute the depth at a point directly
+    /// (no sampling or interpolation).
+    ///
+    /// This method solves for (x, y) that when input into the
+    /// Gerstner wave function
+    /// gives the coordinates of the supplied parameter
+    /// _point (_point.x(), _point.y()),
+    /// and also computes the wave height pz at this point.
+    /// The depth h = pz - point.z().
+    /// This is a numerical method that uses a multi-variate
+    /// Newton solver to solve
+    /// the two dimensional non-linear system. In general it is not as fast as
+    /// sampling from a discretised wave field with an efficient
+    /// line intersection algorithm.
+    ///
+    /// \param[in] _point       The point at which we want the depth.
+    /// \param[in] _time        The time at which we want the depth.
+    /// \param[in] _timeInit    The time at which we want the wavefield to start
+    /// \return                 The depth 'h' at the point.
+    public: double ComputeDepthDirectly(const gz::math::Vector3d &_point,
+                                        double _time,
+                                        double _timeInit = 0);
+
+    /// \brief Get all the parameters in a single message.
+    /// \return A message with all the parameters.
+    public: gz::msgs::Param Parameters() const;
+
+    /// \internal
+    /// \brief Pointer to the class private data.
+    private: std::unique_ptr<WavefieldPrivate> data;
```

Comment:
use `GZ_UTILS_UNIQUE_IMPL_PTR` ?

Reply:
Used in [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)